EIPS/eip-3369.md

```diff
@@ -0,0 +1,183 @@
+func generateDatasetItem(cache []uint32, index uint32, keccak512 hasher) []byte {
+    // Calculate the number of theoretical rows (we use one buffer nonetheless)
+    rows := uint32(len(cache) / hashWords)
+
+    // Initialize the mix
+    mix := make([]byte, hashBytes)
+
+    binary.LittleEndian.PutUint32(mix, cache[(index%rows)*hashWords]^index)
+    for i := 1; i < hashWords; i++ {
+        binary.LittleEndian.PutUint32(mix[i*4:], cache[(index%rows)*hashWords+uint32(i)])
+    }
+    keccak512(mix, mix)
+
+    // Convert the mix to uint32s to avoid constant bit shifting
+    intMix := make([]uint32, hashWords)
+    for i := 0; i < len(intMix); i++ {
+        intMix[i] = binary.LittleEndian.Uint32(mix[i*4:])
+    }
+    // fnv it with a lot of random cache nodes based on index
+    for i := uint32(0); i < datasetParents; i++ {
+        parent := fnv0(index^i, intMix[i%16]) % rows
+        fnvHash(intMix, cache[parent*hashWords:], fnv0)
+    }
+    // Flatten the uint32 mix into a binary one and return
+    for i, val := range intMix {
+        binary.LittleEndian.PutUint32(mix[i*4:], val)
+    }
+    keccak512(mix, mix)
+    return mix
+}
+
+// hashimoto aggregates data from the full dataset in order to produce our final
+// value for a particular header hash and nonce.
+func hashimoto(hash []byte, nonce uint64, size uint64, lookup func(index uint32) []uint32) ([]byte, []byte) {
+    // Calculate the number of theoretical rows (we use one buffer nonetheless)
+    rows := uint32(size / mixBytes)
+
+    // Combine header+nonce into a 64 byte seed
+    seed := make([]byte, 40)
+    copy(seed, hash)
+    binary.LittleEndian.PutUint64(seed[32:], nonce)
+
+    seed = crypto.Keccak512(seed)
+    seedHead := binary.LittleEndian.Uint32(seed)
+
+    // Start the mix with replicated seed
+    mix := make([]uint32, mixBytes/4)
+    for i := 0; i < len(mix); i++ {
+        mix[i] = binary.LittleEndian.Uint32(seed[i%16*4:])
+    }
+    // Mix in random dataset nodes
+    temp := make([]uint32, len(mix))
+
+    for i := 0; i < loopAccesses; i++ {
+        parent := fnvA(uint32(i)^seedHead, mix[i%len(mix)]) % rows
+        for j := uint32(0); j < mixBytes/hashBytes; j++ {
+            copy(temp[j*hashWords:], lookup(2*parent+j))
+        }
+        fnvHash(mix, temp, fnvB)
+    }
+    // Compress mix
+    for i := 0; i < len(mix); i += 4 {
+        mix[i/4] = fnvC(fnvD(fnvE(mix[i], mix[i+1]), mix[i+2]), mix[i+3])
+    }
+    mix = mix[:len(mix)/4]
+
+    digest := make([]byte, common.HashLength)
+    for i, val := range mix {
+        binary.LittleEndian.PutUint32(digest[i*4:], val)
+    }
+    return digest, crypto.Keccak256(append(seed, digest...))
+}
+```
+
+
+# Implementations
+
+- [Go](https://github.com/mineruniter969/ethash1.1-go)
+- [C (Python/JS wrapped)](https://github.com/mineruniter969/ethash1.1-c)
+- Ethminer - TBD
+
+# Test Vectors
+
+See [mineruniter969/ethash1.1-tests](https://github.com/mineruniter969/ethash1.1-tests)
```

Comment:
```suggestion
{
  "first": {
    "nonce": "4242424242424242",
    "mixHash": "aa6a928db9b548ebf20fc9a74e9200321426f1c2db1571636cdd3a33eb162b36",
    "header": "f901f3a00000000000000000000000000000000000000000000000000000000000000000a01dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347940000000000000000000000000000000000000000a09178d0f23c965d81f0834a4c72c6253ce6830f4022b1359aaebfc1ecba442d4ea056e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421a056e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421b90100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000008302000080830f4240808080a058f759ede17a706c93f13030328bcea40c1d1341fb26f2facd21ceb0dae57017884242424242424242",
    "seed": "0000000000000000000000000000000000000000000000000000000000000000",
    "result": "3972318778d2af9d3c5c3dfc463bc2a5ebeebd1a7a04392708ff94d29aa18c5f",
    "cache_size": 16776896,
    "full_size": 1073739904,
    "header_hash": "2a8de2adf89af77358250bf908bf04ba94a6e8c3ba87775564a41d269a05e4ce",
    "cache_hash": "35ded12eecf2ce2e8da2e15c06d463aae9b84cb2530a00b932e4bbc484cde353"
  },
  "second": {
    "nonce": "307692cf71b12f6d",
    "mixHash": "4a2ef8287dc21f5def0d4e9694208c56e574b1692d7b254822a3f4704d8ad1ba",
    "header": "f901f7a01bef91439a3e070a6586851c11e6fd79bbbea074b2b836727b8e75c7d4a6b698a01dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d4934794ea3cb5f94fa2ddd52ec6dd6eb75cf824f4058ca1a00c6e51346be0670ce63ac5f05324e27d20b180146269c5aab844d09a2b108c64a056e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421a056e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421b90100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000008302004002832fefd880845511ed2a80a0e55d02c555a7969361cf74a9ec6211d8c14e4517930a00442f171bdb1698d17588307692cf71b12f6d",
    "seed": "0000000000000000000000000000000000000000000000000000000000000000",
    "result": "5ab98957ba5520d4e367080f442e37a047cfd9d2857b6e00dd12d82900d108a6",
    "cache_size": 16776896,
    "full_size": 1073739904,
    "header_hash": "100cbec5e5ef82991290d0d93d758f19082e71f234cf479192a8b94df6da6bfe",
    "cache_hash": "35ded12eecf2ce2e8da2e15c06d463aae9b84cb2530a00b932e4bbc484cde353"
  }
}
```
Test vectors should either be inlined into the EIP or included as a file in `../assets/eip-3372/<filename>`.  It looks like the test vectors are pretty small, so I recommend just inlining them here.